src/ng/animate.js

```diff
@@ -416,8 +416,8 @@ var $AnimateProvider = ['$provide', /** @this */ function($provide) {
        * @param {Promise} animationPromise The animation promise that is returned when an animation is started.
        */
       cancel: function(runner) {
-        if (runner.end) {
-          runner.end();
+        if (runner.cancel) {
+          runner.cancel();
```

Comment:
Nice detective work :smiley:

From what I understand though, the same failures that the test are demonstrating can happen in real apps, so it is a bug that we should fix. I.e. it is possible that the `ngSwitch` expression changes more than once as part of the same digest loop (which means before `$$rAF` has been called to perform the leave animations). 

In that case, we will hit the same problem of trying to access a value by index, while something else had modified the array by splicing elements.

What I think we should do, is pass the value itself (e.g. the animation runner) and look it up in the `previousLeaveAnimations`, instead of lokking it up by index. I.e. change [ngSwitch.js#L155-L159][1] to:

```diff
-var spliceFactory = function(array, index) {
+var spliceFactory = function(array, item) {
   return function(response) {
-    if (response !== false) array.splice(index, 1);
+    if (response !== false) arrayRemove(array, item);
   };
 };
```

And then change how we call it in [ngSwitch.js#L173][2]:

```diff
-runner.done(spliceFactory(previousLeaveAnimations, i));
+runner.done(spliceFactory(previousLeaveAnimations, runner));
```

This will have a very slight overhead, but considering that the `previousLeaveAnimations` array will be tiny (most of the time just one signle element), it will be negligible.

<sub>I wonder if a similar issue affects other structural directives, e.g. `ngInclude`...</sub>

[1]: https://github.com/Narretz/angular.js/blob/6eae31c8b906e17a2773825be86a32f33bcd5482/src/ng/directive/ngSwitch.js#L155-L159
[2]: https://github.com/Narretz/angular.js/blob/6eae31c8b906e17a2773825be86a32f33bcd5482/src/ng/directive/ngSwitch.js#L173